File: November 2024 Leetcode Solution/November-21.cpp
class Solution {
public:
    int countUnguarded(int m, int n, vector<vector<int>>& guards, vector<vector<int>>& walls) {
        vector<int> grid(m * n, 0); 
        auto index = [&](int x, int y) { return x * n + y; }; 

        for (const auto& guard : guards) grid[index(guard[0], guard[1])] = 2;
        for (const auto& wall : walls) grid[index(wall[0], wall[1])] = 2;

        const int dirs[5] = {-1, 0, 1, 0, -1};

        for (const auto& guard : guards) {
            for (int d = 0; d < 4; ++d) {
                int x = guard[0], y = guard[1];
                int dx = dirs[d], dy = dirs[d + 1];
                while (true) {
                    x += dx;
                    y += dy;

                    if (x < 0 || x >= m || y < 0 || y >= n || grid[index(x, y)] == 2) break;

                    if (grid[index(x, y)] == 0) grid[index(x, y)] = 1;
                }
            }
        }

        return count(grid.begin(), grid.end(), 0);
    }
};

File: November 2024 Leetcode Solution/November-21.java
class Solution {
    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {

        int[][] grid = new int[m][n];

        for (int[] guard : guards) grid[guard[0]][guard[1]] = 2;
        for (int[] wall : walls) grid[wall[0]][wall[1]] = 2;

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int[] guard : guards) {
            for (int[] dir : dirs) {
                int x = guard[0] + dir[0], y = guard[1] + dir[1];
                while (x >= 0 && x < m && y >= 0 && y < n && grid[x][y] != 2) {
                    if (grid[x][y] == 0) grid[x][y] = 1; 
                    x += dir[0];
                    y += dir[1];
                }
            }
        }

        int unguardedCount = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 0) unguardedCount++;
            }
        }

        return unguardedCount;
    }
}
